### PR TITLE
SSCS-11709 - Add error for unavailable dates on update to case data and update other party data

### PR DESCRIPTION
**Pull request by hajam09**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/SscsHelper.java [hunk 1]
```diff
@@ -54,4 +54,41 @@ public static void updateDirectionDueDateByAnAmountOfDays(SscsCaseData sscsCaseD
...
+            }
+        }
+    }
+
+    public static void validateHearingOptionsAndExcludeDates(PreSubmitCallbackResponse<SscsCaseData> response, HearingOptions hearingOptions) {
+        List<String> listOfErrors = new ArrayList<>();
+        List<ExcludeDate> excludeDates = Optional.ofNullable(hearingOptions.getExcludeDates()).orElse(Collections.emptyList());
+
+        if (excludeDates.isEmpty()) {
+            // when the user select yes to unavailable dates, but does not add an entry.
```
Comment (nickhill111) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/SscsHelper.java:84]: remove comment

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/caseupdated/CaseUpdatedAboutToSubmitHandlerTest.java [hunk 2]
```diff
@@ -679,6 +679,89 @@ public void givenHearingTypeOralAndWantsToAttendHearingNo_thenAddWarningMessage(
...
+                        .start("2023-06-17")
+                        .end("2023-06-19")
+                        .build()).build()));
+
+        callback.getCaseDetails().getCaseData().getAppeal().setHearingOptions(HearingOptions.builder().wantsToAttend("Yes").scheduleHearing("Yes").build());
+        callback.getCaseDetails().getCaseData().getAppeal().getHearingOptions().setExcludeDates(excludeDate);
+
+
+        PreSubmitCallbackResponse<SscsCaseData> response = handler.handle(ABOUT_TO_SUBMIT, callback, USER_AUTHORISATION);
+        assertThat(response.getErrors().isEmpty(), is(true));
```
Comment (nickhill111) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/caseupdated/CaseUpdatedAboutToSubmitHandlerTest.java:762]: missing a few bits of CC:
https://sonarcloud.io/summary/new_code?id=SSCSTCA&pullRequest=3335

**Review by nickhill111 — CHANGES_REQUESTED** [review]

**Review by nickhill111 — APPROVED** [review]

**Review by StevensonSi — APPROVED** [review]

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/caseupdated/CaseUpdatedAboutToSubmitHandlerTest.java [hunk 3]
```diff
@@ -679,6 +679,98 @@ public void givenHearingTypeOralAndWantsToAttendHearingNo_thenAddWarningMessage(
...
 
+    @Test
+    public void givenAnAppealWithIncorrectExcludedDateStartDateAfterEndDate_thenProvideErrorMessage() {
+        List<ExcludeDate> excludeDate = new ArrayList<>(List.of(
+                ExcludeDate.builder().value(DateRange.builder()
+                        .start("2023-06-17")
+                        .end("2023-05-18")
+                        .build()).build()));
+
+        callback.getCaseDetails().getCaseData().getAppeal().setHearingOptions(HearingOptions.builder().wantsToAttend("Yes").scheduleHearing("Yes").build());
```
Comment (jarekPierchala) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/caseupdated/CaseUpdatedAboutToSubmitHandlerTest.java:690]: callback.getCaseDetails().getCaseData().getAppeal() to sepatate var

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/caseupdated/CaseUpdatedAboutToSubmitHandlerTest.java [hunk 4]
```diff
@@ -679,6 +679,98 @@ public void givenHearingTypeOralAndWantsToAttendHearingNo_thenAddWarningMessage(
...
+
+    @Test
+    public void givenAnAppealWithIncorrectExcludedDatesStartDateEmpty_thenProvideErrorMessage() {
+        List<ExcludeDate> excludeDate = new ArrayList<>(List.of(
+                ExcludeDate.builder().value(DateRange.builder()
+                        .start(null)
+                        .end("2023-05-17")
+                        .build()).build()));
+
+        callback.getCaseDetails().getCaseData().getAppeal().setHearingOptions(HearingOptions.builder().wantsToAttend("Yes").scheduleHearing("Yes").build());
```
Comment (jarekPierchala) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/caseupdated/CaseUpdatedAboutToSubmitHandlerTest.java:706]: callback.getCaseDetails().getCaseData().getAppeal() to sepatate var

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/caseupdated/CaseUpdatedAboutToSubmitHandlerTest.java [hunk 5]
```diff
@@ -679,6 +679,98 @@ public void givenHearingTypeOralAndWantsToAttendHearingNo_thenAddWarningMessage(
...
+    @Test
+    public void givenAnAppealWithIncorrectExcludedEndDate_thenProvideErrorMessage() {
+        List<ExcludeDate> excludeDate;
+        excludeDate = new ArrayList<>(List.of(
+                ExcludeDate.builder().value(DateRange.builder()
+                        .start("2023-06-17")
+                        .end(null)
+                        .build()).build()));
+
+        callback.getCaseDetails().getCaseData().getAppeal().setHearingOptions(HearingOptions.builder().wantsToAttend("Yes").scheduleHearing("Yes").build());
```
Comment (jarekPierchala) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/caseupdated/CaseUpdatedAboutToSubmitHandlerTest.java:723]: same

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/caseupdated/CaseUpdatedAboutToSubmitHandlerTest.java [hunk 6]
```diff
@@ -679,6 +679,98 @@ public void givenHearingTypeOralAndWantsToAttendHearingNo_thenAddWarningMessage(
...
+    @Test
+    public void givenAnAppealWithCorrectExcludedDates_thenDontProvideError() {
+        List<ExcludeDate> excludeDate;
+        excludeDate = new ArrayList<>(List.of(
+                ExcludeDate.builder().value(DateRange.builder()
+                        .start("2023-06-17")
+                        .end("2023-06-19")
+                        .build()).build()));
+
+        callback.getCaseDetails().getCaseData().getAppeal().setHearingOptions(HearingOptions.builder().wantsToAttend("Yes").scheduleHearing("Yes").build());
```
Comment (jarekPierchala) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/caseupdated/CaseUpdatedAboutToSubmitHandlerTest.java:757]: same

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/updateotherparty/UpdateOtherPartyAboutToSubmitHandlerTest.java [hunk 7]
```diff
@@ -440,6 +441,168 @@ public void givenSscs1CaseOtherPartyWantsToAttendYes_thenHearingTypeNotChangedAn
...
+        otherParty.getValue().setRole(Role.builder().name("ReceivingParent").build());
+
+        ExcludeDate excludeDate1 = ExcludeDate.builder().value(DateRange.builder().start("").end("").build()).build();
+        ExcludeDate excludeDate2 = ExcludeDate.builder().value(DateRange.builder().start(null).end(null).build()).build();
+        ExcludeDate excludeDate3 = ExcludeDate.builder().value(DateRange.builder().start("2023-01-01").end("2023-01-02").build()).build();
+
+        otherParty.getValue().getHearingOptions().setExcludeDates(List.of(excludeDate1, excludeDate2, excludeDate3));
+
+        Appeal appeal = Appeal.builder()
+            .benefitType(BenefitType.builder().code(Benefit.CHILD_SUPPORT.getShortName()).build())
```
Comment (jarekPierchala) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/updateotherparty/UpdateOtherPartyAboutToSubmitHandlerTest.java:492]: BenefitType.builder().code(Benefit.CHILD_SUPPORT.getShortName()).build() as separate var

Comment (jarekPierchala) [reply src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/updateotherparty/UpdateOtherPartyAboutToSubmitHandlerTest.java:492]: Check for simillar stuff below

**Review by jarekPierchala — CHANGES_REQUESTED** [review]

**Review by AndreiTrushkov — DISMISSED** [review]

**Review by jarekPierchala — DISMISSED** [review]

**Review by mrmsali — DISMISSED** [review]

**Review by StevensonSi — APPROVED** [review]

**Review by nickhill111 — DISMISSED** [review]

**Review by nickhill111 — DISMISSED** [review]

**Review by StevensonSi — APPROVED** [review]

**Review by StevensonSi — APPROVED** [review]

**Review by mrmsali — APPROVED** [review]

**Pull request merged**